### PR TITLE
Handle unmatching remote cluster wildcards properly for IndicesRequest.SingleIndexNoWildcards requests

### PR DESCRIPTION
**Pull request by quux00**
When security is enabled and a IndicesRequest.SingleIndexNoWildcards request uses a
wildcard for remote clusters that does not match (for example "*:myindex" when there
are no configured remotes), RBACEngine#authorizeIndexAction throws an assert error
because it is given an empty list of indices.

This fix checks the IndicesRequest.SingleIndexNoWildcards special case in IndicesAndAliasesResolver.resolveIndicesAndAliasesWithoutWildcards and if 
splitLocalAndRemoteIndexNames filters out all indexes due to a non-matching
remote cluster wildcard, that code now just throws the NoSuchRemoteClusterException 
(which is what RBACEngine does when asserts are disabled).

This handles painless/execute requests to remote clusters.

### CODE REVIEW

**Comment by quux00** [comment]
The "quick test" for this is to show that what failed before (when the bug was reported) is now working:

```
./gradlew run
```

In another terminal:

```
curl -s -D /dev/stderr -w "\n" -u elastic-admin:elastic-password -H "Content-Type: application/json" -X POST localhost:9200/_scripts/painless/_execute --data '{"script":{"source":"doc[\"field\"].value.length() <= params.max_length","params":{"max_length":4}},"context":"filter","context_setup":{"index":"*:test","document":{"field":"four"}}}'

HTTP/1.1 404 Not Found
X-elastic-product: Elasticsearch
content-type: application/json
content-length: 203

{"error":{"root_cause":[{"type":"no_such_remote_cluster_exception","reason":"no such remote cluster: [*]"}],"type":"no_such_remote_cluster_exception","reason":"no such remote cluster: [*]"},"status":404}
```

On main, this results in a stack trace and the `./gradlew run` based process exits.

<details>
  <summary><b>Stack trace on main</b> (toggle to view)</summary>

```
[2024-05-29T15:54:30,941][WARN ][o.e.h.AbstractHttpServerTransport] [runTask-0] caught exception while handling client http traffic, closing connection Netty4HttpChannel{localAddress=/[0:0:0:0:0:0:0:1]:9200, remoteAddress=/[0:0:0:0:0:0:0:1]:56794} java.lang.Exception: java.lang.AssertionError: every indices request needs to have its indices set thus the resolved indices must not be empty
	at org.elasticsearch.transport.netty4@8.15.0-SNAPSHOT/org.elasticsearch.http.netty4.Netty4HttpPipeliningHandler.exceptionCaught(Netty4HttpPipeliningHandler.java:447)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:346)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:447)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.codec@4.1.109.Final/io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.codec@4.1.109.Final/io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.codec@4.1.109.Final/io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.codec@4.1.109.Final/io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at org.elasticsearch.transport.netty4@8.15.0-SNAPSHOT/org.elasticsearch.http.netty4.Netty4HttpHeaderValidator.forwardData(Netty4HttpHeaderValidator.java:209)
	at org.elasticsearch.transport.netty4@8.15.0-SNAPSHOT/org.elasticsearch.http.netty4.Netty4HttpHeaderValidator.forwardFullRequest(Netty4HttpHeaderValidator.java:152)
	at org.elasticsearch.transport.netty4@8.15.0-SNAPSHOT/org.elasticsearch.http.netty4.Netty4HttpHeaderValidator$1.lambda$onResponse$0(Netty4HttpHeaderValidator.java:125)
	at io.netty.common@4.1.109.Final/io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.common@4.1.109.Final/io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.common@4.1.109.Final/io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.transport@4.1.109.Final/io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
	at io.netty.common@4.1.109.Final/io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.common@4.1.109.Final/io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: java.lang.AssertionError: every indices request needs to have its indices set thus the resolved indices must not be empty
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.RBACEngine.lambda$authorizeIndexAction$3(RBACEngine.java:390)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:245)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.support.SubscribableListener$SuccessResult.complete(SubscribableListener.java:382)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.support.SubscribableListener.tryComplete(SubscribableListener.java:302)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.support.SubscribableListener.addListener(SubscribableListener.java:205)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.support.SubscribableListener.addListener(SubscribableListener.java:170)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$CachingAsyncSupplier.getAsync(AuthorizationService.java:1079)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.RBACEngine.authorizeIndexAction(RBACEngine.java:389)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.authorizeAction(AuthorizationService.java:510)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.maybeAuthorizeRunAs(AuthorizationService.java:442)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.lambda$authorize$3(AuthorizationService.java:329)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:248)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:32)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.RBACEngine.lambda$resolveAuthorizationInfo$0(RBACEngine.java:155)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:245)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.lambda$getRoles$4(CompositeRolesStore.java:192)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:245)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.lambda$getRole$5(CompositeRolesStore.java:210)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:245)
	at org.elasticsearch.xcore@8.15.0-SNAPSHOT/org.elasticsearch.xpack.core.security.authz.store.RoleReferenceIntersection.lambda$buildRole$0(RoleReferenceIntersection.java:49)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:245)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.support.GroupedActionListener.onResponse(GroupedActionListener.java:56)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:248)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.lambda$buildThenMaybeCacheRole$11(CompositeRolesStore.java:345)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:245)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.lambda$buildRoleFromDescriptors$23(CompositeRolesStore.java:540)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:245)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.NativePrivilegeStore.getPrivileges(NativePrivilegeStore.java:149)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.buildRoleFromDescriptors(CompositeRolesStore.java:532)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.buildThenMaybeCacheRole(CompositeRolesStore.java:321)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.lambda$buildRoleFromRoleReference$8(CompositeRolesStore.java:279)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:248)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.RoleDescriptorStore.lambda$resolveRoleNames$4(RoleDescriptorStore.java:206)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:248)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.RoleDescriptorStore.lambda$loadRoleDescriptorsAsync$9(RoleDescriptorStore.java:268)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:248)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:32)
	at org.elasticsearch.xcore@8.15.0-SNAPSHOT/org.elasticsearch.xpack.core.common.IteratingActionListener.onResponse(IteratingActionListener.java:141)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.RoleDescriptorStore.lambda$loadRoleDescriptorsAsync$13(RoleDescriptorStore.java:295)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:248)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.FileRolesStore.accept(FileRolesStore.java:120)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.FileRolesStore.accept(FileRolesStore.java:65)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.RoleDescriptorStore.lambda$loadRoleDescriptorsAsync$14(RoleDescriptorStore.java:276)
	at org.elasticsearch.xcore@8.15.0-SNAPSHOT/org.elasticsearch.xpack.core.common.IteratingActionListener.onResponse(IteratingActionListener.java:135)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.RoleDescriptorStore.lambda$loadRoleDescriptorsAsync$13(RoleDescriptorStore.java:295)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:248)
	at org.elasticsearch.xcore@8.15.0-SNAPSHOT/org.elasticsearch.xpack.core.security.authz.store.ReservedRolesStore.accept(ReservedRolesStore.java:994)
	at org.elasticsearch.xcore@8.15.0-SNAPSHOT/org.elasticsearch.xpack.core.security.authz.store.ReservedRolesStore.accept(ReservedRolesStore.java:44)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.RoleDescriptorStore.lambda$loadRoleDescriptorsAsync$14(RoleDescriptorStore.java:276)
	at org.elasticsearch.xcore@8.15.0-SNAPSHOT/org.elasticsearch.xpack.core.common.IteratingActionListener.run(IteratingActionListener.java:117)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.RoleDescriptorStore.loadRoleDescriptorsAsync(RoleDescriptorStore.java:297)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.RoleDescriptorStore.roleDescriptors(RoleDescriptorStore.java:243)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.RoleDescriptorStore.resolveRoleNames(RoleDescriptorStore.java:186)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.RoleDescriptorStore.resolveNamedRoleReference(RoleDescriptorStore.java:90)
	at org.elasticsearch.xcore@8.15.0-SNAPSHOT/org.elasticsearch.xpack.core.security.authz.store.RoleReference$NamedRoleReference.resolve(RoleReference.java:71)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.buildRoleFromRoleReference(CompositeRolesStore.java:273)
	at org.elasticsearch.xcore@8.15.0-SNAPSHOT/org.elasticsearch.xpack.core.security.authz.store.RoleReferenceIntersection.lambda$buildRole$1(RoleReferenceIntersection.java:53)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at org.elasticsearch.xcore@8.15.0-SNAPSHOT/org.elasticsearch.xpack.core.security.authz.store.RoleReferenceIntersection.buildRole(RoleReferenceIntersection.java:53)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.getRole(CompositeRolesStore.java:208)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.getRoles(CompositeRolesStore.java:185)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.RBACEngine.resolveAuthorizationInfo(RBACEngine.java:151)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.authorize(AuthorizationService.java:345)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.lambda$applyInternal$5(SecurityActionFilter.java:178)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:245)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:95)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authc.AuthenticatorChain.authenticate(AuthenticatorChain.java:93)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:264)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:173)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.applyInternal(SecurityActionFilter.java:174)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.apply(SecurityActionFilter.java:131)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:93)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:68)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.tasks.TaskManager.registerAndExecute(TaskManager.java:196)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.client.internal.node.NodeClient.executeLocally(NodeClient.java:105)
	at org.elasticsearch.painless@8.15.0-SNAPSHOT/org.elasticsearch.painless.action.PainlessExecuteAction$RestAction.lambda$prepareRequest$0(PainlessExecuteAction.java:875)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.rest.BaseRestHandler.handleRequest(BaseRestHandler.java:119)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.rest.RestController$1.onResponse(RestController.java:478)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.rest.RestController$1.onResponse(RestController.java:472)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.rest.SecurityRestFilter.doHandleRequest(SecurityRestFilter.java:89)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.rest.SecurityRestFilter.lambda$intercept$0(SecurityRestFilter.java:81)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:248)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authc.support.SecondaryAuthenticator.lambda$authenticateAndAttachToContext$3(SecondaryAuthenticator.java:99)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:245)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authc.support.SecondaryAuthenticator.authenticate(SecondaryAuthenticator.java:109)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.authc.support.SecondaryAuthenticator.authenticateAndAttachToContext(SecondaryAuthenticator.java:90)
	at org.elasticsearch.security@8.15.0-SNAPSHOT/org.elasticsearch.xpack.security.rest.SecurityRestFilter.intercept(SecurityRestFilter.java:75)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.rest.RestController.dispatchRequest(RestController.java:472)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.rest.RestController.tryAllHandlers(RestController.java:632)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.rest.RestController.dispatchRequest(RestController.java:329)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.http.AbstractHttpServerTransport.dispatchRequest(AbstractHttpServerTransport.java:487)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.http.AbstractHttpServerTransport.handleIncomingRequest(AbstractHttpServerTransport.java:583)
	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.http.AbstractHttpServerTransport.incomingRequest(AbstractHttpServerTransport.java:460)
	at org.elasticsearch.transport.netty4@8.15.0-SNAPSHOT/org.elasticsearch.http.netty4.Netty4HttpPipeliningHandler.handlePipelinedRequest(Netty4HttpPipeliningHandler.java:127)
	at org.elasticsearch.transport.netty4@8.15.0-SNAPSHOT/org.elasticsearch.http.netty4.Netty4HttpPipeliningHandler.channelRead(Netty4HttpPipeliningHandler.java:117)
	at io.netty.transport@4.1.109.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	... 25 more
```

</details>

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java [hunk 1]
```diff
@@ -164,6 +165,18 @@ ResolvedIndices resolveIndicesAndAliasesWithoutWildcards(String action, IndicesR
...
+            if (split.getLocal().isEmpty() && split.getRemote().isEmpty()) {
+                for (String indexExpression : indices) {
+                    String[] clusterAndIndex = indexExpression.split(":", 2);
+                    if (clusterAndIndex.length == 2) {
+                        if (clusterAndIndex[0].contains("*")) {
+                            throw new NoSuchRemoteClusterException(clusterAndIndex[0]);
+                        }
+                    }
+                }
+            }
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java:179]: I see that `NoSuchRemoteClusterException` is used in code for concrete remote cluster names (not cluster name wildcards).

Also, in many other places we just use `IndexNotFoundException`s, so I would prefer this simpler version:
```
diff --git a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java
index 42a1d89a9aa..99359089e37 100644
--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java
@@ -167,15 +167,8 @@ class IndicesAndAliasesResolver {
             split = remoteClusterResolver.splitLocalAndRemoteIndexNames(indicesRequest.indices());
             // all indices can come back empty when the remote index expression included a cluster alias with a wildcard
             // and no remote clusters are configured that match it
-            if (split.getLocal().isEmpty() && split.getRemote().isEmpty()) {
-                for (String indexExpression : indices) {
-                    String[] clusterAndIndex = indexExpression.split(":", 2);
-                    if (clusterAndIndex.length == 2) {
-                        if (clusterAndIndex[0].contains("*")) {
-                            throw new NoSuchRemoteClusterException(clusterAndIndex[0]);
-                        }
-                    }
-                }
+            if (split.isEmpty()) {
+                throw new IndexNotFoundException(Arrays.toString(indicesRequest.indices()));
             }
         } else {
             split = new ResolvedIndices(Arrays.asList(indicesRequest.indices()), List.of());
``` 

But, more importantly, what is the behavior when Security is disabled, when `_scripts/painless/_execute` goes for a remote cluster name wildcard that doesn't cover any configured cluster names? I think this is the error we should also throw here.

**Review by albertzaharovits — COMMENTED** [review]
Thanks for fixing!
I asked a clarifying question on what's the response error in case security is disabled (on the querying cluster), because it's important to have consistent behaviors when enabling security.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java (hunk 1 above)
Comment (quux00) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java:179]: > But, more importantly, what is the behavior when Security is disabled, when _scripts/painless/_execute goes for a remote cluster name wildcard that doesn't cover any configured cluster names?

When security is disabled and a wildcard is used that does not match a remote cluster, `NoSuchRemoteClusterException` is thrown:

```
 curl -s -D /dev/stderr -w "\n" -H "Content-Type: application/json" -X POST localhost:9200/_scripts/painless/_execute --data '{"script":{"source":"doc[\"field\"].value.length() <= params.max_length","params":{"max_length":4}},"context":"filter","context_setup":{"index":"w*w:test","document":{"field":"four"}}}'

HTTP/1.1 404 Not Found
X-elastic-product: Elasticsearch
content-type: application/json
content-length: 207

{"error":{"root_cause":[{"type":"no_such_remote_cluster_exception","reason":"no such remote cluster: [w*w]"}],"type":"no_such_remote_cluster_exception","reason":"no such remote cluster: [w*w]"},"status":404}
```

(I tested the above on both `main` and this branch to make sure I didn't change that behavior as well).

So with this PR, both routes (security enabled or disabled) return the same error message.

In the latest push, I have updated the painless API docs to indicate that wildcards are not allowed in either the cluster alias or the index portion of the index expression and list what error to expect. See [this commit](https://github.com/elastic/elasticsearch/pull/109185/commits/d0d913bf8035a0c72ec4ac8b3730e9146d8344b).

**Review by albertzaharovits — APPROVED** [review]

**Pull request merged**